File: tests/rosetta/transpiler/php/find-chess960-starting-position-identifier-2.php
<?php
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function _indexof($s, $sub) {
    $pos = strpos($s, $sub);
    return $pos === false ? -1 : $pos;
}
$__start_mem = memory_get_usage();
$__start = _now();
  $g2lMap = ['♜' => 'R', '♞' => 'N', '♝' => 'B', '♛' => 'Q', '♚' => 'K', '♖' => 'R', '♘' => 'N', '♗' => 'B', '♕' => 'Q', '♔' => 'K'];
  $names = ['R' => 'rook', 'N' => 'knight', 'B' => 'bishop', 'Q' => 'queen', 'K' => 'king'];
  $ntable = ['01' => 0, '02' => 1, '03' => 2, '04' => 3, '12' => 4, '13' => 5, '14' => 6, '23' => 7, '24' => 8, '34' => 9];
  function indexOf($s, $sub) {
  global $g2lMap, $names, $ntable;
  $i = 0;
  while ($i <= strlen($s) - strlen($sub)) {
  if (substr($s, $i, $i + strlen($sub) - $i) == $sub) {
  return $i;
}
  $i = $i + 1;
};
  return -1;
};
  function removeChar($s, $ch) {
  global $g2lMap, $names, $ntable;
  $res = '';
  $i = 0;
  while ($i < strlen($s)) {
  $c = substr($s, $i, $i + 1 - $i);
  if ($c != $ch) {
  $res = $res . $c;
}
  $i = $i + 1;
};
  return $res;
};
  function g2l($pieces) {
  global $g2lMap, $names, $ntable;
  $res = '';
  $i = 0;
  while ($i < strlen($pieces)) {
  $ch = substr($pieces, $i, $i + 1 - $i);
  $res = $res . $g2lMap[$ch];
  $i = $i + 1;
};
  return $res;
};
  function countChar($s, $ch) {
  global $g2lMap, $names, $ntable;
  $c = 0;
  $i = 0;
  while ($i < strlen($s)) {
  if (substr($s, $i, $i + 1 - $i) == $ch) {
  $c = $c + 1;
}
  $i = $i + 1;
};
  return $c;
};
  function spid($pieces) {
  global $g2lMap, $names, $ntable;
  $pieces = g2l($pieces);
  if (strlen($pieces) != 8) {
  return -1;
}
  foreach (['K', 'Q'] as $one) {
  if (countChar($pieces, $one) != 1) {
  return -1;
}
};
  foreach (['R', 'N', 'B'] as $two) {
  if (countChar($pieces, $two) != 2) {
  return -1;
}
};
  $r1 = _indexof($pieces, 'R');
  $r2 = _indexof(substr($pieces, $r1 + 1, strlen($pieces) - ($r1 + 1)), 'R') + $r1 + 1;
  $k = _indexof($pieces, 'K');
  if ($k < $r1 || $k > $r2) {
  return -1;
}
  $b1 = _indexof($pieces, 'B');
  $b2 = _indexof(substr($pieces, $b1 + 1, strlen($pieces) - ($b1 + 1)), 'B') + $b1 + 1;
  if (($b2 - $b1) % 2 == 0) {
  return -1;
}
  $piecesN = removeChar(removeChar($pieces, 'Q'), 'B');
  $n1 = _indexof($piecesN, 'N');
  $n2 = _indexof(substr($piecesN, $n1 + 1, strlen($piecesN) - ($n1 + 1)), 'N') + $n1 + 1;
  $N = $ntable[_str($n1) . _str($n2)];
  $piecesQ = removeChar($pieces, 'B');
  $Q = _indexof($piecesQ, 'Q');
  $D = _indexof('0246', _str($b1));
  $L = _indexof('1357', _str($b2));
  if ($D == (0 - 1)) {
  $D = _indexof('0246', _str($b2));
  $L = _indexof('1357', _str($b1));
}
  return 96 * $N + 16 * $Q + 4 * $D + $L;
};
  function main() {
  global $g2lMap, $names, $ntable;
  foreach (['♕♘♖♗♗♘♔♖', '♖♘♗♕♔♗♘♖', '♖♕♘♗♗♔♖♘', '♖♘♕♗♗♔♖♘'] as $pieces) {
  echo rtrim($pieces . ' or ' . g2l($pieces) . ' has SP-ID of ' . _str(spid($pieces))), PHP_EOL;
};
};
  main();
$__end = _now();
$__end_mem = memory_get_usage();
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;

File: tests/rosetta/transpiler/FS/find-chess960-starting-position-identifier-2.fs
// Generated 2025-07-30 21:05 +0700

exception Return

let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _substring (s:string) (start:int) (finish:int) =
    let len = String.length s
    let mutable st = if start < 0 then len + start else start
    let mutable en = if finish < 0 then len + finish else finish
    if st < 0 then st <- 0
    if st > len then st <- len
    if en > len then en <- len
    if st > en then st <- en
    s.Substring(st, en - st)

let mutable g2lMap: Map<string, string> = Map.ofList [("♜", "R"); ("♞", "N"); ("♝", "B"); ("♛", "Q"); ("♚", "K"); ("♖", "R"); ("♘", "N"); ("♗", "B"); ("♕", "Q"); ("♔", "K")]
let mutable names: Map<string, string> = Map.ofList [("R", "rook"); ("N", "knight"); ("B", "bishop"); ("Q", "queen"); ("K", "king")]
let mutable ntable: Map<string, int> = Map.ofList [("01", 0); ("02", 1); ("03", 2); ("04", 3); ("12", 4); ("13", 5); ("14", 6); ("23", 7); ("24", 8); ("34", 9)]
let rec indexOf (s: string) (sub: string) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable s = s
    let mutable sub = sub
    try
        let mutable i: int = 0
        while i <= ((String.length s) - (String.length sub)) do
            if (_substring s i (i + (String.length sub))) = sub then
                __ret <- i
                raise Return
            i <- i + 1
        __ret <- -1
        raise Return
        __ret
    with
        | Return -> __ret
and removeChar (s: string) (ch: string) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable s = s
    let mutable ch = ch
    try
        let mutable res: string = ""
        let mutable i: int = 0
        while i < (String.length s) do
            let c: string = _substring s i (i + 1)
            if c <> ch then
                res <- res + c
            i <- i + 1
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
and g2l (pieces: string) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable pieces = pieces
    try
        let mutable res: string = ""
        let mutable i: int = 0
        while i < (String.length pieces) do
            let ch: string = _substring pieces i (i + 1)
            res <- res + (unbox<string> (g2lMap.[ch] |> unbox<string>))
            i <- i + 1
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
and countChar (s: string) (ch: string) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable s = s
    let mutable ch = ch
    try
        let mutable c: int = 0
        let mutable i: int = 0
        while i < (String.length s) do
            if (_substring s i (i + 1)) = ch then
                c <- c + 1
            i <- i + 1
        __ret <- c
        raise Return
        __ret
    with
        | Return -> __ret
and spid (pieces: string) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable pieces = pieces
    try
        pieces <- g2l pieces
        if (String.length pieces) <> 8 then
            __ret <- -1
            raise Return
        for one in [|"K"; "Q"|] do
            if (int (countChar pieces (unbox<string> one))) <> 1 then
                __ret <- -1
                raise Return
        for two in [|"R"; "N"; "B"|] do
            if (int (countChar pieces (unbox<string> two))) <> 2 then
                __ret <- -1
                raise Return
        let r1: int = pieces.IndexOf("R")
        let r2 = (int ((int ((_substring pieces (r1 + 1) (String.length pieces)).IndexOf("R"))) + r1)) + 1
        let k: int = pieces.IndexOf("K")
        if (k < r1) || (k > (int r2)) then
            __ret <- -1
            raise Return
        let b1: int = pieces.IndexOf("B")
        let b2 = (int ((int ((_substring pieces (b1 + 1) (String.length pieces)).IndexOf("B"))) + b1)) + 1
        if (int (((((int b2) - b1) % 2 + 2) % 2))) = 0 then
            __ret <- -1
            raise Return
        let mutable piecesN: string = removeChar (removeChar pieces "Q") "B"
        let n1: int = piecesN.IndexOf("N")
        let n2 = (int ((int ((_substring piecesN (n1 + 1) (String.length piecesN)).IndexOf("N"))) + n1)) + 1
        let N: int = ntable.[((string n1) + (string n2))] |> unbox<int>
        let mutable piecesQ: string = removeChar pieces "B"
        let Q: int = piecesQ.IndexOf("Q")
        let mutable D: int = "0246".IndexOf(string b1)
        let mutable L: int = "1357".IndexOf(string b2)
        if D = (0 - 1) then
            D <- int ("0246".IndexOf(string b2))
            L <- int ("1357".IndexOf(string b1))
        __ret <- (((96 * N) + (16 * Q)) + (4 * D)) + L
        raise Return
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        let __bench_start = _now()
        let __mem_start = System.GC.GetTotalMemory(true)
        for pieces in [|"♕♘♖♗♗♘♔♖"; "♖♘♗♕♔♗♘♖"; "♖♕♘♗♗♔♖♘"; "♖♘♕♗♗♔♖♘"|] do
            printfn "%s" (((((unbox<string> pieces) + " or ") + (unbox<string> (g2l (unbox<string> pieces)))) + " has SP-ID of ") + (string (spid (unbox<string> pieces))))
        let __bench_end = _now()
        let __mem_end = System.GC.GetTotalMemory(true)
        printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

        __ret
    with
        | Return -> __ret
main()
